include/boost/url/experimental/detail/router.hpp

```diff
@@ -0,0 +1,205 @@
+//
+// Copyright (c) 2022 Alan de Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/url
+//
+
+#ifndef BOOST_URL_EXPERIMENTAL_DETAIL_ROUTER_HPP
+#define BOOST_URL_EXPERIMENTAL_DETAIL_ROUTER_HPP
+
+#include <boost/url/detail/config.hpp>
+#include <boost/url/decode_view.hpp>
+#include <boost/url/variant.hpp>
+#include <boost/url/string_view.hpp>
+#include <boost/url/grammar/all_chars.hpp>
+#include <boost/url/grammar/alnum_chars.hpp>
+#include <boost/url/grammar/alpha_chars.hpp>
+#include <boost/url/grammar/lut_chars.hpp>
+#include <boost/url/grammar/optional_rule.hpp>
+#include <boost/url/grammar/token_rule.hpp>
+#include <boost/url/grammar/variant_rule.hpp>
+#include <boost/url/grammar/not_empty_rule.hpp>
+#include <boost/url/grammar/detail/charset.hpp>
+#include <boost/url/rfc/pchars.hpp>
+#include <boost/url/rfc/detail/path_rules.hpp>
+#include <boost/core/ignore_unused.hpp>
+#include <vector>
+#include <locale>
+
+namespace boost {
+namespace urls {
+namespace experimental {
+namespace detail {
+
+struct match_any_t {};
+
+// A dynamic path segment
+class segment_template
+{
+    std::string str_;
+    bool is_literal_ = true;
+
+    friend struct segment_template_rule_t;
+public:
+    segment_template() = default;
+
+    bool
+    match(pct_string_view seg) const
+    {
+        if (is_literal_)
+            return *seg == str_;
+
+        return true;
+    }
+
+    string_view
+    string() const
+    {
+        return str_;
+    }
+
+
+    string_view
+    id() const
+    {
+        if (is_literal_)
+            return {};
+        return string_view{str_}
+            .substr(1, str_.find_first_of(":}") - 1);
+    }
+
+    bool
+    empty() const
+    {
+        return str_.empty();
+    }
+
+    bool
+    is_literal() const
+    {
+        return is_literal_;
+    }
+
+    friend
+    bool operator==(
+        segment_template const& a,
+        segment_template const& b)
+    {
+        if (a.is_literal_ != b.is_literal_)
+            return false;
+        if (a.is_literal_)
+            return a.str_ == b.str_;
+        return true;
+    }
+};
+
+// The syntax of dynamic url components is a
+// subset of:
+// https://fmt.dev/latest/syntax.html
+
+// digit             ::=  "0"..."9"
+constexpr auto integer_chars =
+    grammar::lut_chars("0123456789");
+
+// identifier        ::=  id_start id_continue*
+// id_start          ::=  "a"..."z" | "A"..."Z" | "_"
+// id_continue       ::=  id_start | digit
+struct identifier_rule_t
```

Comment:
you shouldn't break it down to this level of granularity... there's no need for an "id rule"